### ISSUE TITLE
No need to use Secret and Access Keys

### ISSUE THREAD

**Issue by zorangagic** [prowler-cloud/prowler#85]
No need to use Secret and Access Keys...Here is how:

http://docs.aws.amazon.com/cli/latest/topic/config-vars.html
The CLI config file on linux is usually in ~/.aws/config, here is an example:
[profile trusting-account]
region = us-east-1 
role_arn = arn:aws:iam::222222222222:role/crossaccount-readonly
external_id = 12345
source_profile = default

**Comment by toniblyx** [comment]
You are right. In that case you only would have to use -p trusting-account, like with any other profile. 

Do you thing I should change something to make it clear? It works that way.

**Comment by zorangagic** [comment]
I would recommend not using  Secret and Access Keys and using roles instead. AWS CLI works exactly the same, just need to create role and update ~/.aws/config.

As prowler is a security tool, we should encourage users to use good security principles and avoid using Secret and Access Keys. You could updated README.md and give them two options:
- Option 1 (preferred ): Using  roles and role_arn in ~/.aws/config
- Option 2: Using  Secret and Access Keys

**Comment by toniblyx** [comment]
Ok, I get your point now. In that case is a user side configuration or hardening check. It might be another extra check that looks at local config file and if any key is found, warn about it. WDYT?

**Comment by MrSecure** [comment]
If I'm reading that AWS doc correctly, there's no need to use the Secret & Access keys when prowler is running on an EC2 instance which was launched with an IAM role.

Is that correct?

**Comment by zorangagic** [comment]
Yes, you can use cross-account roles instead of keys. I do this with all of my accounts...works fine. Once cross- account roles are created just add them into ~/.aws/config and then use ./prowler  -p trusting-account

**Comment by unk1nd0n3** [comment]
How to rotate then external id (aka password or salt) in Cross-account role? How to monitor this? Programmatic keys has this option to show you age of secrets

**Comment by zorangagic** [comment]
To rotate cross-account roles:
1. create second cross-account role with different id
2. add second role as new profile to  ~/.aws/config
3. test application
4. rename profile created in step 2 in ~/.aws/config to old profile and delete old (original) profile
5. delete old cross-account role 

Yet cross-account roles are more secure as it much harder for credentials to leak due to:
- can only be used with trusting account (as opposed to keys that can be used anywhere)
AND
- someone has to know  external id to use it

**Comment by toniblyx** [comment]
I agree with you guys @zorangagic and @MrSecure. I can add a new extra check (extra74) to look for API access keys in ~/.aws/config, regardless if it is an EC2 instance or a workstation. If a key pair is found, show WARNING and link of the official documentation. It is a good check to suggest people to make their admin tasks more secure. 

Something like this? (it would check all profiles regardless the -p option):
```
./prowler -p profile_x -c extra74
7.4  Ensure there are no AWS API access keys listed in your local ~/.aws/config file (Scored)
       WARNING! Profile profile_x has AWS API access key pair listed in your local config file.
       INFO! Use cross-account roles or instance profile if you are running prowler in an EC2 instance. More info: http://docs.aws.amazon.com/cli/latest/topic/config-vars.html
```
Or should it be just INFO? What do you think?

Btw @SrebniukNik I don't get your point. What check do you suggest to add?

**Comment by unk1nd0n3** [comment]
Hi,

on my opinion its not a super secure variant to replace secret key with roles:
1. Role description is still present in config file and can be leaked in a same way as a secret key
2. AWS has not builtin mechanism to verify the age of a role (this is needed for regular credential rotation)
3. Role attached to AWS hosts is potential vector of infrastructure compromise. 

Its better to use these one:
1. Use MFA in conjunction with role or secret keys. See this https://docs.aws.amazon.com/cli/latest/userguide/cli-roles.html
2. Use aws-keychain (if macOS) to store in keychain keys and secrets.

**Comment by toniblyx** [comment]
Good point @SrebniukNik. In that case, the most secure solution is to use assume role with MFA. We could add a command to parse aws/config and find proper configuration with roles and MFA. Something like this:
```
[profile marketingadmin]
role_arn = arn:aws:iam::123456789012:role/marketingadmin
source_profile = default
mfa_serial = arn:aws:iam::123456789012:mfa/jonsmith
```
or this:
```
[profile crossaccountrole]
role_arn = arn:aws:iam::234567890123:role/xaccount
source_profile = default
mfa_serial = arn:aws:iam::123456789012:mfa/jonsmith
external_id = 123456
```

Should be the recommended / more secure configuration. Do you guys agree?

**Comment by toniblyx** [comment]
This issue should be solved by the PR #150. In case I missed something, please feel free to open it again.
